packages/react/src/components/primitive/root.tsx

```diff
@@ -0,0 +1,315 @@
+   */
+  dropzone?: {
+    isDragActive: boolean;
+  };
+};
+
+const PrimitiveContext = createContext<PrimitiveContextValues | null>(null);
+
+export function PrimitiveContextMergeProvider({
+  value,
+  ...props
+}: ProviderProps<Partial<PrimitiveContextValues>>) {
+  const currentValue = useContext(PrimitiveContext);
+
+  if (currentValue === null) {
+    throw new Error(
+      "<PrimitiveContextMergeProvider> must be used within a <UT.Root>",
+    );
+  }
+
+  return (
+    <PrimitiveContext.Provider
+      value={{ ...currentValue, ...value }}
+      {...props}
+    />
+  );
+}
+
+export function usePrimitiveValues(componentName?: string) {
+  const values = useContext(PrimitiveContext);
+  if (values === null) {
+    const name = componentName ? `<UT.${componentName}>` : "usePrimitiveValues";
+    throw new Error(`${name} must be used within a <UT.Root>`);
+  }
+  return values;
+}
+
+export function PrimitiveSlot({
+  children,
+  componentName,
+  default: defaultChildren,
+}: {
+  children: PrimitiveComponentChildren;
+  componentName?: string;
+  default?: React.ReactNode;
+}) {
+  if (!children) return defaultChildren;
+  return typeof children === "function"
+    ? children?.(usePrimitiveValues(componentName))
+    : children;
+}
+
+export type HasDisplayName = {
+  displayName: string;
+};
+
+export type RefProp<T extends (...args: any[]) => any> = T extends (
+  props: any,
+  ref: Ref<infer RefType>,
+) => any
+  ? { ref?: Ref<RefType> }
+  : never;
+
+export type PrimitiveComponentProps<TTag extends ElementType = "div"> = Omit<
+  React.ComponentPropsWithoutRef<TTag>,
+  "children"
+> &
+  PrimitiveComponentChildrenProp & {
+    as?: TTag;
+  };
+
+export type PrimitiveComponentChildrenProp = {
+  children?: PrimitiveComponentChildren;
+};
+
+export type PrimitiveComponentChildren =
+  | ((values: PrimitiveContextValues) => React.ReactNode)
+  | React.ReactNode;
+
+/** These are some internal stuff we use to test the component and for forcing a state in docs */
+type UploadThingInternalProps = {
+  __internal_state?: "readying" | "ready" | "uploading";
+  __internal_upload_progress?: number;
+  __internal_button_disabled?: boolean;
+};
+
+export type RootPrimitiveComponentProps<
+  TRouter extends FileRouter,
+  TEndpoint extends keyof TRouter,
+> = UploadthingComponentProps<TRouter, TEndpoint> & {
+  // TODO: add @see comment for docs
+  children?: PrimitiveComponentChildren;
+  files?: File[];
+  onFilesChange?: (_: File[]) => void;
+};
+
+export function Root<
+  TRouter extends FileRouter,
+  TEndpoint extends keyof TRouter,
+>(
+  props: FileRouter extends TRouter
+    ? ErrorMessage<"You forgot to pass the generic">
+    : RootPrimitiveComponentProps<TRouter, TEndpoint>,
+) {
+  // Cast back to UploadthingComponentProps<TRouter> to get the correct type
+  // since the ErrorMessage messes it up otherwise
+  const $props = props as unknown as RootPrimitiveComponentProps<
```

Comment:
this component will never be used standalone right? it will only be used as part of the set of generated components? 

If so, we can skip all these type guards and nasty casting

```suggestion
export function Root<
  TRouter extends FileRouter,
  TEndpoint extends keyof TRouter,
>(props: RootPrimitiveComponentProps<TRouter, TEndpoint>) {
  props;
```
The conditional was just a dx-enhancement since people could use the components directly from "@uploadthing/react" before and this helped catch if they didn't bind the generic